tls/autotls-client.md

```diff
@@ -0,0 +1,396 @@
+	4. Node sends a POST request to `/v1/_acme-challenge` endpoint using `payload` as HTTP body and `headers` as HTTP headers.
+	5. Node SHOULD save the `bearer` token from the `authentication-info` response header, and use it for following requests to the AutoTLS broker.
+
+
+
+## Signalling challenge completion to ACME server
+1. Node SHOULD query DNS records (`TXT _acme-challenge.{b36peerid}.libp2p.direct` and `A dashed-public-ip-address.{b36peerid}.libp2p.direct`) until they are set by the AutoTLS broker.
+
+**Note:** Here, `dashed-public-ip-address` is the public IPv4 address on in which the node received the confirmation dial from the broker.
+For example, if the node has two public IPv4 addresses `1.1.1.1` and `8.8.8.8`, and the broker dialed it through `1.1.1.1`, then the node SHOULD query the `A 1-1-1-1.{b36peerid}.libp2p.direct`.
+
+**Note:** The node SHOULD NOT send more than `max_dns_retries` DNS requests.
+After `max_dns_timeout`, the communication is considered failed.
+What to do after `max_dns_timeout` has passed is left as an implementation decision.
+
+2. Node notifies the ACME server about challenge completion so that the ACME server can lookup the DNS resource records that the AutoTLS broker has set. The notification is done in the form of a POST request to `chalUrl` with an empty HTTP body (`{}`).
+	1. Node sends an empty signed JSON payload (`{}`) to the ACME server using the `kid` obtained from the initial ACME registration and gets the response from the server (`completedResponse`).
+	2. Node extracts `url` field from `completedResponse`'s JSON body. The extracted URL is named `checkUrl` in this document.
+3. The node polls the ACME server by sending a GET HTTP request to `checkUrl` with an empty body, and sign using the `kid` of the registered account. The node MUST poll the ACME server until it receives a response with `status: valid` or `status: invalid` field, meaning that the challenge checking was successful or not, respectively.
+
+**Note:** The node SHOULD NOT send more than `max_acme_poll_retries` poll requests to the ACME server.
+After `max_acme_timeout`, the communication has failed.
+What to do after `max_acme_timeout` has passed is left as an implementation decision.
+
+
+
+## Downloading certificate
+1. Node finalizes the certificate request:
+    1. Generate CSR for the `*.{b36peerid}.libp2p.direct` domain.
+    2. Encode the CSR with URL safe base 64 (`b64CSR`).
+    3. Send a `kid` signed POST request to `finalizeUrl` with JSON HTTP body of `{"csr": b64CSR}`.
+2. Node MUST poll ACME server by sending GET requests to `orderUrl` until the ACME server's response contains a `status` field with a value different than `processing`.
+
+**Note:** The node SHOULD NOT send more than `max_acme_poll_retries` poll requests to the ACME server.
+After `max_acme_timeout`, the communication has failed.
+What to do after `max_acme_timeout` has passed is left as an implementation decision.
```

Comment:
Similar ask: spec should clearly note that there should be exponential backoff for retry attempts.